File: problems/arrays-and-hashing/3-valid-anagram/cpp/main.cpp
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

bool isAnagram(string s, string t) {
    unsigned sSize = s.size();
    unsigned tSize = t.size();
    if (sSize != tSize) {
        return false;
    }
    unordered_map<char, unsigned> sHashMap;
    unordered_map<char, unsigned> tHashMap;
    for (int i = 0; i < sSize; ++i) {
        sHashMap[s[i]] += 1;
        tHashMap[t[i]] += 1;
    }

    return sHashMap == tHashMap;
}

int main() {
    string s1 = "racecar";
    string t1 = "carrace";
    string s2 = "jar";
    string t2 = "jam";

    bool res1 = isAnagram(s1, t1);
    bool res2 = isAnagram(s2, t2);

    if (res1 != true && res2 != false) {
        cout << "Actual results not equal with expected results." << endl;
    }
    else {
        cout << "Results correct." << endl;
    }

    return 0;
}

File: problems/arrays-and-hashing/3-valid-anagram/rust/src/main.rs
use std::collections::HashMap;

fn is_anagram(s: String, t: String) -> bool {
    let s_chars: Vec<char> = s.chars().collect();
    let t_chars: Vec<char> = t.chars().collect();
    let s_size = s.len();
    let t_size = t.len();
    if s_size != t_size {
        return false
    }
    let mut s_hashmap = HashMap::new();
    let mut t_hashmap = HashMap::new();
    for i in 0..s_size {
        *s_hashmap.entry(s_chars[i]).or_insert(0) += 1;
        *t_hashmap.entry(t_chars[i]).or_insert(0) += 1;
    }

    return s_hashmap == t_hashmap
}

fn main() {
    let s_1   = String::from("racecar");
    let t_1 = String::from("carrace");
    let s_2   = String::from("jar");
    let t_2 = String::from("jam");

    let res_1 = is_anagram(s_1, t_1);
    let res_2 = is_anagram(s_2, t_2);

    if res_1 != true && res_2 != false {
        println!("Actual results not equal with expected results.");
    }
    else {
        println!("Results correct.");
    }
}
